## Install requirements

In [1]:
!pip install transformers
!pip install datasets
!pip install --upgrade pandas
!pip install evaluate

DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
spyder 4.1.5 requires pyqt5<5.13; python_version >= "3", which is not installed.
spyder 4.1.5 requires pyqtwebengine<5.13; python_version >= "3", which is not installed.



   ---------------------------------------- 10.8/10.8 MB 12.3 MB/s eta 0:00:00
   --------------------------------------- 229.9/229.9 kB 14.6 MB/s eta 0:00:00
   --------------------------------------- 345.4/345.4 kB 10.8 MB/s eta 0:00:00
  Attempting uninstall: python-dateutil
    Found existing installation: python-dateutil 2.8.1
    Uninstalling python-dateutil-2.8.1:
      Successfully uninstalled python-dateutil-2.8.1
  Attempting uninstall: pandas
    Found existing installation: pandas 1.1.3
    Uninstalling pandas-1.1.3:
      Successfully uninstalled pandas-1.1.3


DEPRECATION: pyodbc 4.0.0-unsupported has a non-standard version number. pip 24.1 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of pyodbc or contact the author to suggest that they release a version with a conforming version number. Discussion can be found at https://github.com/pypa/pip/issues/12063


In [48]:
import torch

In [96]:

text_column_name = "text" #@param {type:"string"}
label_column_name = "demon" #@param {type:"string"}

model_name = "distilbert-base-uncased" #@param {type:"string"}
test_size = 0.2 #@param {type:"number"}
num_labels = 2 #@param {type:"number"}

device = 'cuda' if torch.cuda.is_available() else 'cpu'

## Read and Prepare the Dataset

In [97]:
import pandas as pd

In [98]:
df = pd.read_csv(r"C:\Users\91971\demon_slayer.csv")

In [99]:
df.head(15)

Unnamed: 0     Character Name Breathing/Demonic Style  \
0            0     Tanjiro Kamado         Water Breathing   
1            1   Zenitsu Agatsuma       Thunder Breathing   
2            2  Inosuke Hashibira         Beast Breathing   
3            3      Kanao Tsuyuri        Flower Breathing   
4            4  Genya Shinazugawa                     NaN   
5            5             Murata         Water Breathing   
6            6              Ozaki                     NaN   
7            7      Nezuko Kamado         Blood Demon Art   
8            8    Muzan Kibutsuji                     NaN   
9            9          Kokushibo          Moon Breathing   
10          10               Doma           Ice Breathing   
11          11              Akaza       Destructive Death   
12          12             Nakime          Biwa Demon Art   
13          13           Hantengu    Emotion Manipulation   
14          14             Gyokko       Fish Manipulation   

                                          Description Human or Demon  
0   Tanjiro Kamado uses Water Breathing techniques...          Human  
1   Zenitsu Agatsuma wields Thunder Breathing, whi...          Human  
2   Inosuke Hashibira developed Beast Breathing on...          Human  
3   Kanao Tsuyuri employs Flower Breathing, which ...          Human  
4   Genya Shinazugawa lacks a specific breathing s...          Human  
5   Murata is a practitioner of Water Breathing, s...          Human  
6   Ozaki is a lesser-known Demon Slayer without a...          Human  
7   Nezuko Kamado uses her Blood Demon Art to enha...          Demon  
8   Muzan Kibutsuji, the Demon King, has no specif...          Demon  
9   Kokushibo uses Moon Breathing, which combines ...          Demon  
10  Doma wields Ice Breathing, creating deadly ice...          Demon  
11  Akaza employs Destructive Death, a combat styl...          Demon  
12  Nakime uses her Biwa Demon Art to manipulate t...          Demon  
13  Hantengu's abilities allow him to manipulate e...          Demon  
14  Gyokko has the ability to manipulate fish and ...          Demon

In [100]:
df = df.fillna("None")

In [101]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33 entries, 0 to 32
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Unnamed: 0               33 non-null     int64 
 1   Character Name           33 non-null     object
 2   Breathing/Demonic Style  33 non-null     object
 3   Description              33 non-null     object
 4   Human or Demon           33 non-null     object
dtypes: int64(1), object(4)
memory usage: 1.4+ KB


In [102]:
df1 = df.iloc[:, [4, 3]]
df1

Human or Demon                                        Description
0           Human  Tanjiro Kamado uses Water Breathing techniques...
1           Human  Zenitsu Agatsuma wields Thunder Breathing, whi...
2           Human  Inosuke Hashibira developed Beast Breathing on...
3           Human  Kanao Tsuyuri employs Flower Breathing, which ...
4           Human  Genya Shinazugawa lacks a specific breathing s...
5           Human  Murata is a practitioner of Water Breathing, s...
6           Human  Ozaki is a lesser-known Demon Slayer without a...
7           Demon  Nezuko Kamado uses her Blood Demon Art to enha...
8           Demon  Muzan Kibutsuji, the Demon King, has no specif...
9           Demon  Kokushibo uses Moon Breathing, which combines ...
10          Demon  Doma wields Ice Breathing, creating deadly ice...
11          Demon  Akaza employs Destructive Death, a combat styl...
12          Demon  Nakime uses her Biwa Demon Art to manipulate t...
13          Demon  Hantengu's abilities allow him to manipulate e...
14          Demon  Gyokko has the ability to manipulate fish and ...
15          Demon  Gyutaro utilizes a Blood Demon Art that enhanc...
16          Demon  Daki also uses a Blood Demon Art that manipula...
17          Demon  Kaigaku, a former Thunder Breathing user, beco...
18          Demon  Enmu's Dream Manipulation allows him to contro...
19          Demon  Rokuro, a lower-ranked demon, relies on his br...
20          Demon  Wakuraba's abilities are not well-documented, ...
21          Demon  Mukago's combat style is not specified, but sh...
22          Demon  Rui manipulates threads to create deadly webs ...
23          Demon  Kamanue's abilities and style are not prominen...
24          Demon  Kyogai uses Drum Manipulation to control his s...
25          Human  Gyomei Himejima uses Stone Breathing and is th...
26          Human  Muichiro Tokito, the Mist Hashira, utilizes Mi...
27          Human  Mitsuri Kanroji employs Love Breathing, a uniq...
28          Human  Obanai Iguro uses Serpent Breathing, character...
29          Human  Sanemi Shinazugawa wields Wind Breathing, maki...
30          Human  Tengen Uzui, the Sound Hashira, uses Sound Bre...
31          Human  Shinobu Kocho uses Insect Breathing, which rel...
32          Human  Kyojuro Rengoku is the Flame Hashira, masterin...

In [103]:
df['text'] = df['Character Name']+'. '+df['Description']
df

Unnamed: 0      Character Name Breathing/Demonic Style  \
0            0      Tanjiro Kamado         Water Breathing   
1            1    Zenitsu Agatsuma       Thunder Breathing   
2            2   Inosuke Hashibira         Beast Breathing   
3            3       Kanao Tsuyuri        Flower Breathing   
4            4   Genya Shinazugawa                    None   
5            5              Murata         Water Breathing   
6            6               Ozaki                    None   
7            7       Nezuko Kamado         Blood Demon Art   
8            8     Muzan Kibutsuji                    None   
9            9           Kokushibo          Moon Breathing   
10          10                Doma           Ice Breathing   
11          11               Akaza       Destructive Death   
12          12              Nakime          Biwa Demon Art   
13          13            Hantengu    Emotion Manipulation   
14          14              Gyokko       Fish Manipulation   
15          15             Gyutaro         Blood Demon Art   
16          16                Daki         Blood Demon Art   
17          17             Kaigaku       Thunder Breathing   
18          18                Enmu      Dream Manipulation   
19          19              Rokuro                    None   
20          20            Wakuraba                    None   
21          21              Mukago                    None   
22          22                 Rui     Thread Manipulation   
23          23             Kamanue                    None   
24          24              Kyogai       Drum Manipulation   
25          25     Gyomei Himejima         Stone Breathing   
26          26     Muichiro Tokito          Mist Breathing   
27          27     Mitsuri Kanroji          Love Breathing   
28          28        Obanai Iguro       Serpent Breathing   
29          29  Sanemi Shinazugawa          Wind Breathing   
30          30         Tengen Uzui         Sound Breathing   
31          31       Shinobu Kocho        Insect Breathing   
32          32     Kyojuro Rengoku         Flame Breathing   

                                          Description Human or Demon  \
0   Tanjiro Kamado uses Water Breathing techniques...          Human   
1   Zenitsu Agatsuma wields Thunder Breathing, whi...          Human   
2   Inosuke Hashibira developed Beast Breathing on...          Human   
3   Kanao Tsuyuri employs Flower Breathing, which ...          Human   
4   Genya Shinazugawa lacks a specific breathing s...          Human   
5   Murata is a practitioner of Water Breathing, s...          Human   
6   Ozaki is a lesser-known Demon Slayer without a...          Human   
7   Nezuko Kamado uses her Blood Demon Art to enha...          Demon   
8   Muzan Kibutsuji, the Demon King, has no specif...          Demon   
9   Kokushibo uses Moon Breathing, which combines ...          Demon   
10  Doma wields Ice Breathing, creating deadly ice...          Demon   
11  Akaza employs Destructive Death, a combat styl...          Demon   
12  Nakime uses her Biwa Demon Art to manipulate t...          Demon   
13  Hantengu's abilities allow him to manipulate e...          Demon   
14  Gyokko has the ability to manipulate fish and ...          Demon   
15  Gyutaro utilizes a Blood Demon Art that enhanc...          Demon   
16  Daki also uses a Blood Demon Art that manipula...          Demon   
17  Kaigaku, a former Thunder Breathing user, beco...          Demon   
18  Enmu's Dream Manipulation allows him to contro...          Demon   
19  Rokuro, a lower-ranked demon, relies on his br...          Demon   
20  Wakuraba's abilities are not well-documented, ...          Demon   
21  Mukago's combat style is not specified, but sh...          Demon   
22  Rui manipulates threads to create deadly webs ...          Demon   
23  Kamanue's abilities and style are not prominen...          Demon   
24  Kyogai uses Drum Manipulation to control his s...          Demon   
25  Gyomei Himejima use

In [104]:
df = df[['text','Human or Demon']]

In [105]:
df = df.dropna()

In [106]:
df

text Human or Demon
0   Tanjiro Kamado. Tanjiro Kamado uses Water Brea...          Human
1   Zenitsu Agatsuma. Zenitsu Agatsuma wields Thun...          Human
2   Inosuke Hashibira. Inosuke Hashibira developed...          Human
3   Kanao Tsuyuri. Kanao Tsuyuri employs Flower Br...          Human
4   Genya Shinazugawa. Genya Shinazugawa lacks a s...          Human
5   Murata. Murata is a practitioner of Water Brea...          Human
6   Ozaki. Ozaki is a lesser-known Demon Slayer wi...          Human
7   Nezuko Kamado. Nezuko Kamado uses her Blood De...          Demon
8   Muzan Kibutsuji. Muzan Kibutsuji, the Demon Ki...          Demon
9   Kokushibo. Kokushibo uses Moon Breathing, whic...          Demon
10  Doma. Doma wields Ice Breathing, creating dead...          Demon
11  Akaza. Akaza employs Destructive Death, a comb...          Demon
12  Nakime. Nakime uses her Biwa Demon Art to mani...          Demon
13  Hantengu. Hantengu's abilities allow him to ma...          Demon
14  Gyokko. Gyokko has the ability to manipulate f...          Demon
15  Gyutaro. Gyutaro utilizes a Blood Demon Art th...          Demon
16  Daki. Daki also uses a Blood Demon Art that ma...          Demon
17  Kaigaku. Kaigaku, a former Thunder Breathing u...          Demon
18  Enmu. Enmu's Dream Manipulation allows him to ...          Demon
19  Rokuro. Rokuro, a lower-ranked demon, relies o...          Demon
20  Wakuraba. Wakuraba's abilities are not well-do...          Demon
21  Mukago. Mukago's combat style is not specified...          Demon
22  Rui. Rui manipulates threads to create deadly ...          Demon
23  Kamanue. Kamanue's abilities and style are not...          Demon
24  Kyogai. Kyogai uses Drum Manipulation to contr...          Demon
25  Gyomei Himejima. Gyomei Himejima uses Stone Br...          Human
26  Muichiro Tokito. Muichiro Tokito, the Mist Has...          Human
27  Mitsuri Kanroji. Mitsuri Kanroji employs Love ...          Human
28  Obanai Iguro. Obanai Iguro uses Serpent Breath...          Human
29  Sanemi Shinazugawa. Sanemi Shinazugawa wields ...          Human
30  Tengen Uzui. Tengen Uzui, the Sound Hashira, u...          Human
31  Shinobu Kocho. Shinobu Kocho uses Insect Breat...          Human
32  Kyojuro Rengoku. Kyojuro Rengoku is the Flame ...          Human

In [107]:
df['Human or Demon'].value_counts()

Human or Demon
Demon    18
Human    15
Name: count, dtype: int64

### Label Encoder

In [108]:
from sklearn import preprocessing

In [109]:
le = preprocessing.LabelEncoder()
le.fit(df['Human or Demon'].tolist())
df['label'] = le.transform(df['Human or Demon'].tolist())

In [110]:
df.head()

text Human or Demon  label
0  Tanjiro Kamado. Tanjiro Kamado uses Water Brea...          Human      1
1  Zenitsu Agatsuma. Zenitsu Agatsuma wields Thun...          Human      1
2  Inosuke Hashibira. Inosuke Hashibira developed...          Human      1
3  Kanao Tsuyuri. Kanao Tsuyuri employs Flower Br...          Human      1
4  Genya Shinazugawa. Genya Shinazugawa lacks a s...          Human      1

# Class weights

In [111]:
from sklearn.utils.class_weight import compute_class_weight

In [112]:
class_weights = compute_class_weight('balanced',
                     classes=sorted(df['label'].unique().tolist()),
                     y=df['label'].tolist()).tolist()

In [113]:
class_weights

[0.9166666666666666, 1.1]

### Train/Test Split

In [114]:
from sklearn.model_selection import train_test_split

In [115]:
df_train,df_test = train_test_split(df,test_size=test_size,stratify=df['label'])

### Convert to Huggingface Dataset

In [116]:
from datasets import Dataset

In [117]:
train_dataset = Dataset.from_pandas(df_train)
test_dataset = Dataset.from_pandas(df_test)

### Tokenizer

In [118]:
from transformers import AutoTokenizer

In [119]:
tokenizer = AutoTokenizer.from_pretrained(model_name)

def preprocess_function(examples):
    return tokenizer(examples["text"], truncation=True)

C:\Users\91971\Anaconda3\lib\site-packages\huggingface_hub\file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


In [120]:
tokenized_train = train_dataset.map(preprocess_function, batched=True)

Map:   0%|          | 0/26 [00:00<?, ? examples/s]

In [121]:
tokenized_test = test_dataset.map(preprocess_function, batched=True)

Map:   0%|          | 0/7 [00:00<?, ? examples/s]

## Initialize Model

In [122]:
from transformers import AutoModelForSequenceClassification

In [123]:
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_labels)

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_layer_norm.weight', 'vocab_transform.weight', 'vocab_layer_norm.bias', 'vocab_projector.weight', 'vocab_projector.bias', 'vocab_transform.bias']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight', 'pre_classifier

## Train model

In [124]:
from transformers import DataCollatorWithPadding
from transformers import TrainingArguments, Trainer
import evaluate
import numpy as np
import torch
from torch import nn

In [125]:
data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

In [126]:
metric = evaluate.load("accuracy")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)

In [127]:
class CustomTrainer(Trainer):
    def compute_loss(self, model, inputs, return_outputs=False):
        labels = inputs.get("labels")
        # forward pass
        outputs = model(**inputs)
        logits = outputs.get("logits")
        # compute custom loss (suppose one has 3 labels with different weights)
        loss_fct = nn.CrossEntropyLoss(weight=torch.tensor(class_weights).to(device=device))
        loss = loss_fct(logits.view(-1, self.model.config.num_labels), labels.view(-1))
        return (loss, outputs) if return_outputs else loss

In [128]:
training_args = TrainingArguments(
    output_dir="./results",
    learning_rate=2e-4,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=5,
    weight_decay=0.01,
    evaluation_strategy = "epoch",
    logging_strategy="epoch"
)

trainer = CustomTrainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_train,
    eval_dataset=tokenized_test,
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
    
)


In [129]:
trainer.train()

C:\Users\91971\Anaconda3\lib\site-packages\transformers\optimization.py:391: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
You're using a DistilBertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


TrainOutput(global_step=20, training_loss=0.27164395824074744, metrics={'train_runtime': 31.3317, 'train_samples_per_second': 4.149, 'train_steps_per_second': 0.638, 'total_flos': 1514060970576.0, 'train_loss': 0.27164395824074744, 'epoch': 5.0})

In [130]:
trainer.save_model('demon_slayer_model')

## Evaluate Model

In [131]:
from sklearn.metrics import classification_report

### On Training set

In [134]:
preds = trainer.predict(tokenized_train)
preds = np.argmax(preds[:2][0],axis=1)
GT = df_train['label'].tolist()
print(classification_report(GT,preds))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      1.00      1.00        12

    accuracy                           1.00        26
   macro avg       1.00      1.00      1.00        26
weighted avg       1.00      1.00      1.00        26



### On Test set

In [135]:
preds = trainer.predict(tokenized_test)
preds = np.argmax(preds[:2][0],axis=1) #preds[:3][1]
GT = df_test['label'].tolist()
print(classification_report(GT,preds))

              precision    recall  f1-score   support

           0       0.75      0.75      0.75         4
           1       0.67      0.67      0.67         3

    accuracy                           0.71         7
   macro avg       0.71      0.71      0.71         7
weighted avg       0.71      0.71      0.71         7



- Precision: Precision is the ratio of correctly predicted positive observations to the total predicted positives. For class 0, it's 0.75, indicating that 75% of the instances predicted as class 0 were correct. For class 1, it's 0.67, indicating that 67% of the instances predicted as class 1 were correct.

- Recall: Recall is the ratio of correctly predicted positive observations to the all observations in actual class. For class 0, it's 0.75, meaning that 75% of the actual class 0 instances were correctly predicted. For class 1, it's 0.67, meaning that 67% of the actual class 1 instances were correctly predicted.

- F1-score: F1-score is the weighted average of Precision and Recall. It considers both false positives and false negatives. For class 0, it's 0.75, and for class 1, it's 0.67.

- Support: Support is the number of actual occurrences of the class in the specified dataset. For class 0, it's 4, and for class 1, it's 3.

- Accuracy: Accuracy is the ratio of correctly predicted observation to the total observations. Here, it's 0.71, meaning 71% of the predictions were correct.